File: solutions/js/day2-step2-issue-token.js
// ============================================================
// Day 2 - Step 2 Solution: Issue a Custom Token
// Learn: How to create a trust line and issue tokens on XRPL.
// The holder must trust the issuer before receiving tokens.
// ============================================================

const xrpl = require('xrpl');
const fs = require('fs');

const TESTNET_URL = 'wss://s.altnet.rippletest.net:51233';

async function main() {
  const { wallets } = JSON.parse(fs.readFileSync('../../wallets.json', 'utf-8'));
  const issuerWallet = xrpl.Wallet.fromSeed(wallets[0].seed);
  const holderWallet = xrpl.Wallet.fromSeed(wallets[1].seed);

  const currencyCode = 'ACD';

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();
  console.log('Connected to XRPL Testnet');

  // Create trust line
  const trustSetTx = {
    TransactionType: 'TrustSet',
    Account: holderWallet.address,
    LimitAmount: {
      currency: currencyCode,
      issuer: issuerWallet.address,
      value: '1000000'
    }
  };
  const trustResult = await client.submitAndWait(trustSetTx, { wallet: holderWallet });
  console.log('Trust line created!', trustResult.result.hash);

  // Issue tokens
  const paymentTx = {
    TransactionType: 'Payment',
    Account: issuerWallet.address,
    Destination: holderWallet.address,
    Amount: {
      currency: currencyCode,
      issuer: issuerWallet.address,
      value: '500'
    }
  };
  const payResult = await client.submitAndWait(paymentTx, { wallet: issuerWallet });
  console.log('Tokens issued!', payResult.result.hash);

  console.log(`\nExplorer: https://testnet.xrpl.org/accounts/${issuerWallet.address}`);
  await client.disconnect();
}

main().catch(console.error);

File: solutions/python/day2_step2_issue_token.py
"""
Day 2 Step 2 Solution: Issue Token
====================================
Create a TrustLine and issue a custom token on XRPL Testnet.
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import TrustSet, Payment
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.transaction import submit_and_wait

TESTNET_URL = 'wss://s.altnet.rippletest.net:51233'

# Load wallets
with open('../../wallets.json') as f:
    data = json.load(f)
    wallets = data['wallets']

issuer = Wallet.from_seed(wallets[0]['seed'])
holder = Wallet.from_seed(wallets[1]['seed'])
currency_code = 'TST'

print(f"Issuer: {issuer.address}")
print(f"Holder: {holder.address}")

with WebsocketClient(TESTNET_URL) as client:

    # TrustSet from holder to issuer
    trust_tx = TrustSet(
        account=holder.address,
        limit_amount=IssuedCurrencyAmount(
            currency=currency_code,
            issuer=issuer.address,
            value='1000000'
        )
    )
    trust_result = submit_and_wait(trust_tx, client, holder)
    print(f"TrustSet: {trust_result.result['meta']['TransactionResult']}")

    # Issue tokens from issuer to holder
    payment_tx = Payment(
        account=issuer.address,
        destination=holder.address,
        amount=IssuedCurrencyAmount(
            currency=currency_code,
            issuer=issuer.address,
            value='500'
        )
    )
    payment_result = submit_and_wait(payment_tx, client, issuer)
    print(f"Payment: {payment_result.result['meta']['TransactionResult']}")

# Explorer: https://testnet.xrpl.org/accounts/{address}
print(f"\nExplorer (issuer): https://testnet.xrpl.org/accounts/{issuer.address}")
print(f"Explorer (holder): https://testnet.xrpl.org/accounts/{holder.address}")
